classes/ThankYous/ThankYousRepository.php

```diff
@@ -0,0 +1,1121 @@
+
+		$query->AddWhereAndClause(self::THANKED_TABLE . ".item_id IN in:int:thank_you_ids", $ids);
+
+		$results = $this->db->query($query->GetQuery());
+
+		$thank_yous_thankeds = [];
+		$thankeds            = [];
+		while ($row = $results->fetchArray())
+		{
+			$id                    = (int) $row['id'];
+			$thank_you_id          = (int) $row['item_id'];
+			$owner_class_id        = (int) $row['object_type'];
+			$owner_classes_item_id = (int) $row['object_id'];
+
+			$thankeds[$id] = ['oclass' => $owner_class_id, 'id' => $owner_classes_item_id];
+
+			$thank_yous_thankeds[$thank_you_id][$id] = true;
+		}
+
+		try
+		{
+			$thankeds = $this->CreateThankablesFromOClasses($thankeds);
+		} catch (ThankYouOClass $exception)
+		{
+			throw new LogicException("Unexpected Exception thrown", null, $exception);
+		}
+
+		foreach ($thank_yous_thankeds as $thank_you_id => $thank_you_thankeds)
+		{
+			foreach ($thank_you_thankeds as $id => $true)
+			{
+				$thank_yous_thankeds[$thank_you_id][$id] = $thankeds[$id];
+			}
+		}
+
+		return $thank_yous_thankeds;
+	}
+
+	/**
+	 * Given and array of Thank You IDs, returns an array of Thank You Users, indexed by the Thank Yous's ID.
+	 *
+	 * @param int[] $ids
+	 * @return array[]
+	 */
+	public function GetThankYousUsersByThankYouIds(array $ids)
+	{
+		if (count($ids) === 0)
+		{
+			return [];
+		}
+
+		$query_string = "SELECT * FROM " . self::THANKED_USERS_TABLE;
+		$query        = $this->query_factory->GetQueryBuilder($query_string);
+
+		$query->AddWhereAndClause(self::THANKED_USERS_TABLE . ".thanks_id IN in:int:ids", $ids);
+
+		$results = $this->db->query($query->GetQuery());
+
+		$user_ids = [];
+		$rows     = [];
+		while ($row = $results->fetchArray())
+		{
+			$thank_you_id = (int) $row['thanks_id'];
+			$user_id      = (int) $row['user_id'];
+
+			$rows[]             = ['thank_you_id' => $thank_you_id, 'user_id' => $user_id];
+			$user_ids[$user_id] = true;
+		}
+
+		$user_ids = array_keys($user_ids);
+
+		$users = $this->GetUsers($user_ids);
+
+		$thank_yous_users = [];
+		foreach ($rows as $row)
+		{
+			if (isset($users[$row['user_id']]))
+			{
+				$thank_yous_users[$row['thank_you_id']][$row['user_id']] = $users[$row['user_id']];
+			}
+		}
+
+		return $thank_yous_users;
+	}
+
+	public function GetTagsTotalThankYouUses(?array $orders = null, ?int $limit = null, ?int $offset = null, ?array $extranet_ids = null, bool $allow_no_thanked = true, ?array $date_range = null, ?array $thanked_user_ids = null, ?array $tag_ids = null)
+	{
+		$order = "";
+		if (isset($orders))
+		{
+			$order = $this->utility->BuildOrderString($orders);
+		}
+
+		$query_string = "SELECT COUNT(" . self::THANK_YOU_TAGS_TABLE . ".item_id) AS \"" . self::THANK_YOU_TAGS_TABLE . ".total_uses\"";
+		$query_string .= ", " . self::TAG_TABLE . ".id AS \"" . self::TAG_TABLE . ".id\"";
+		$query_string .= " FROM " . self::TAG_TABLE;
+		$query_string .= $order;
+		$query_string .= " GROUP BY " . self::TAG_TABLE . ".id";
```

Comment:
The identifier quotes in the query string should be removed.

Reply:
I don't understand why this is preferable to leaving them in?
Either way, fixed in f14c164b23f404f37b8286e8da61e4416fedd5c0

Reply:
Reverted in b0534719ee2bc984fd280fc7371aac24c56482e4 as this caused a crash in the query due to the `.` in the aliases.

Reply:
Ah, I didn't realise it was just for aliasing. This is fine then. 👌